Tutorial for identifying a user using the Google OAuth2/OpenID endpoint (#1757)
GoogleChrome/developer.chrome.com: site/en/docs/webstore/identify_user/index.md

description: How to get the Google Account identity of a Chrome Web Store user.
---

> Supporting Google account sign-in can help provide a better user experience. Users of the Chrome Web Store are likely to be logged in to their Google account already, so they won't have to set up and remember yet another username and password.
> 
> This tutorial builds an extension using the [Google OAuth2/OpenID][google-openid] endpoint and the [Chrome
> Identity API][identity-api]. When the user clicks the [action][action], the extension will launch a consent
> screen. After the user signs in to their Google account, the background will log their
> information in the console.

Reviewer: ```suggestion
Supporting Google account sign-in can help provide a better user experience. Users of the Chrome Web
Store are likely to be logged in to their Google account already, so they won't have to set up and
remember yet another username and password.

This tutorial builds an extension using the [Google OAuth2/OpenID][google-openid] endpoint and the
[Chrome Identity API][identity-api]. When the user clicks the [action][action], the extension will
launch a consent screen. After the user signs in to their Google account, the background will log
their information in the console.
```

Rewrap at 100 characters.

To identify the user with Google OAuth2/OpenId, follow these steps:

1. TBD

> 2. TBD

Author: These steps will match the headers. I wanted to ask Simeon his thoughts on the headers before I add the steps 🙂

1. Create your extension files
2. Keep a consistent extension ID
3. Get the OAuth client ID
4. Launch the Authorization flow
5. View the user information

@dotproto What are your thoughts on the headers? 🤔

> We'll go into detail about each step below.

Author: Good point!

## Create your extension files {: #create-extension-files}

Begin by creating a directory and the following starter files.

### manifest.json {: #manifest}

Add the manifest by creating a file called `manifest.json` and include the following code:

```json
{
  "name": "Google OpenID Connect Example",
  "version": "1.0",
  "description": "Use OpenID Connect to identify the user",
  "manifest_version": 3,
  "action": {
    "default_title": "Sign In with Google Accounts"
  },
  "background": {
    "service_worker": "background.js"
  }
}
```

### background.js {: #background}

> Add the background service worker by creating a file called `background.js`. Include the following code:

Reviewer: ```suggestion
Add the background service worker by creating a file called `background.js`. Include the following
code:
```

Rewrap at 100 characters.

```javascript
// background.js

chrome.action.onClicked.addListener(function () {
  console.log('action clicked')
})
```

{% include 'partials/extensions/reusing-prod-extension-id.md' %}

## Get the OAuth client ID {: #get-client-id}

> Navigate to the [Google API Console][google-console] and create a new project. To get a OAuth client ID, follow these steps:

Reviewer: ```suggestion
Navigate to the [Google API Console][google-console] and create a new project. To get a OAuth client
ID, follow these steps:
```

Rewrap at 100 characters.

**1. Customize the consent screen and select scope.**
   - Click the **OAuth consent screen** menu item.
   - Fill out the required consent screen information.
   - Add the **openid** scope.
   - Add **Test users**.
   - Click **Save > Continue**.
  
**2. Create credentials.**   
   - Click the **Credentials** menu item.
   - Click **Create new credentials > Auth Client ID**.
   - Select Application type **Web application**.
   - Enter the **Name** of the OAuth2 client
   - Add `https://<YOUR_EXTENSION_ID>.chromiumapp.org/` as the **Authorized redirect URI**.
   - Finish by clicking **Create**. 

The console will provide an OAuth client ID. Keep this ID for later use.

## Launch the Authorization flow {: #auth-flow}

### Request the "identity" permission {: #identity-permission}

> To use the [Chrome Identity API][identity-api], declare the `"identity"` permission in the `manifest.json`.

Reviewer: ```suggestion
To use the [Chrome Identity API][identity-api], declare the `"identity"` permission in the
`manifest.json`.
```

Rewrap at 100 characters.

```json
{
  "name": "Google OpenID Connect example",
  ...
  "permissions": [
    "identity"
  ],
  ...
}
```

### Construct the authorization URL {: #auth-url}

Before the extension can make a request to Google’s OAuth2 endpoint using the Identity API,

> we need to construct an [authorization URL][auth-url]. It must contain several request parameters

Author: I actually debated about this for a while 😅, thanks for clearing this up for me!

> Before the extension can make a request to Google’s OAuth2 endpoint using the Identity API,
> you need to construct an [authorization URL][auth-url]. It must contain several request parameters
> including the client ID and redirect URI. In addition to the `openid` [scope][openid-scopes], you can request
> `profile` information and/or `email`. Update `background.js` to match the following code:

Reviewer: ```suggestion
Before the extension can make a request to Google’s OAuth2 endpoint using the Identity API, you need
to construct an [authorization URL][auth-url]. It must contain several request parameters including
the client ID and redirect URI. In addition to the `openid` [scope][openid-scopes], you can request
`profile` information and/or `email`. Update `background.js` to match the following code:
```

Rewrap at 100 characters.

> `profile` information and/or `email`. Update `background.js` to match the following code.

Reviewer 2: ```suggestion
`profile` information and/or `email`. Update `background.js` to match the following code:
```

```javascript
// background.js

let clientId = '<CLIENT_ID>'
let redirectUri = `https://${chrome.runtime.id}.chromiumapp.org/`
let nonce = Math.random().toString(36).substring(2, 15)

chrome.action.onClicked.addListener(function () {
  const authUrl = new URL('https://accounts.google.com/o/oauth2/v2/auth')

  authUrl.searchParams.set('client_id', clientId)
  authUrl.searchParams.set('response_type', 'id_token')
  authUrl.searchParams.set('redirect_uri', redirectUri)
  // Add the OpenID scope. Scopes allow you to access the user’s information.
  authUrl.searchParams.set('scope', 'openid profile email')
  authUrl.searchParams.set('nonce', nonce)
  // Show the consent screen after login
  authUrl.searchParams.set('prompt', 'consent')

})

```

> Replace <CLIENT_ID> with the API key generated from the Google console. 

Reviewer 2: ```suggestion
Replace `<CLIENT_ID>` with the API key generated from the Google console. 
```

### Retrieve a redirect URL {: #redirect-url}

Now that the extension has the client ID, redirect URI, and OAuth URL, it can initiate Google's

> authentication flow. Use the [`identity.launchWebAuthFlow`][identity-webauthflow] method to launch the web auth flow and

Reviewer 2: ```suggestion
authentication flow. Call [`identity.launchWebAuthFlow()`][identity-webauthflow] to launch the web auth flow and
```

> Now that the extension has the client ID, redirect URI, and OAuth URL, it can initiate Google's
> authentication flow. Call [`identity.launchWebAuthFlow()`][identity-webauthflow] to launch the web auth flow and
> retrieve a redirect URL. 

Reviewer: ```suggestion
Now that the extension has the client ID, redirect URI, and OAuth URL, it can initiate Google's
authentication flow. Call [`identity.launchWebAuthFlow()`][identity-webauthflow] to launch the web
auth flow and retrieve a redirect URL. 
```

Rewrap at 100 characters.

The redirect URL contains a JSON Web Token (JWT) that identifies the user. To view the requested
user identity information, you'll need to parse the JWT into a plain JavaScript object. Update

> `background.js` to match the following code.

Reviewer 2: ```suggestion
`background.js` to match the following code:
```

```javascript

> // background.js
> ...
> 
> chrome.action.onClicked.addListener(function () {
> 
> ...
> 
>   chrome.identity.launchWebAuthFlow(
>       {
>         url: authUrl.href,
>         interactive: true,
>       },
>       (redirectUrl) => {
>         if (redirectUrl) {
>           // The ID token is in the URL hash
>           const urlHash = redirectUrl.split('#')[1]
>           const params = new URLSearchParams(urlHash)
>           const jwt = params.get('id_token')
> 
>           // Parse the JSON Web Token
>           const base64Url = jwt.split('.')[1]
>           const base64 = base64Url.replace('-', '+').replace('_', '/')
>           const token = JSON.parse(atob(base64))
> 
>           console.log('token', token)
>         }
>       },
>     )
> })

Reviewer: ```suggestion
// background.js

...

chrome.action.onClicked.addListener(function() {

  ...

  chrome.identity.launchWebAuthFlow(
      {
        url: authUrl.href,
        interactive: true,
      },
      (redirectUrl) => {
        if (redirectUrl) {
          // The ID token is in the URL hash
          const urlHash = redirectUrl.split('#')[1];
          const params = new URLSearchParams(urlHash);
          const jwt = params.get('id_token');

          // Parse the JSON Web Token
          const base64Url = jwt.split('.')[1];
          const base64 = base64Url.replace('-', '+').replace('_', '/');
          const token = JSON.parse(atob(base64));

          console.log('token', token);
        }
      },
    );
});
```

Tweaks to better fit Google's JS style guide. 

* Remove extra newline per [4.1.2 Nonempty blocks: K&R style](https://google.github.io/styleguide/jsguide.html#formatting-nonempty-blocks).
* Add semicolons per [4.3.2 Semicolons are required](https://google.github.io/styleguide/jsguide.html#formatting-semicolons-are-required).
* Remove space between function and parenthese per [4.6.2 Horizontal whitespace](https://google.github.io/styleguide/jsguide.html#formatting-horizontal-whitespace).

```

{% Aside %}

The above code is **not** production ready. We strongly encourage validating and decoding the JWT before

> the information it contains is trusted. See [how to handle credential
> responses][credential-responses]

Reviewer 2: ```suggestion
the information it contains is trusted. For more information, see [how to handle credential
responses][credential-responses].
```

Author: It flows better, thanks!

> The above code is **not** production ready. We strongly encourage validating and decoding the JWT before
> the information it contains is trusted. For more information, see [how to handle credential
> responses][credential-responses].

Reviewer: ```suggestion
The above code is **not** production ready. We strongly encourage validating and decoding the JWT
before the information it contains is trusted. For more information, see [how to handle credential
responses][credential-responses].
```

Rewrap at 100 characters.

{% endAside %}

## View the user information {: #user-info}

> Reload and return to the extension. Click the extension action button to start the web authentication flow. Sign in with your Google
> Account, then press Enter. 

Reviewer: ```suggestion
Reload and return to the extension. Click the extension action button to start the web
authentication flow. Sign in with your Google Account, then press Enter. 
```

Rewrap at 100 characters.

{% Img src="image/BhuKGJaIeLNPW9ehns59NfwqKxF2/CETlvMvFpe23QyIAq8Lx.png", alt="ALT_TEXT_HERE",
width="358", height="428" %}

> Go to the extension management page. Select the **service worker**  blue link next to

Reviewer 2: ```suggestion
Go to the extension management page. Select the **service worker** blue link next to
```
